### ISSUE TITLE
Explore possibility of generic types

### ISSUE THREAD

**Issue by AndrewIngram** [facebook/graphql#190]
projects like Relay have shown, it's relatively common to repeat the same generic structures of types multiple times within a project. In the case of Relay, I'm talking about Connections.

The GraphQL definition language already has explicit support for one particular form of generic type, arrays:

```
type Foo {
   id: ID!
   bars: [Bar]
}
```

I'd like to start discussion about being able to do something similar for user-defined structures:

```
generic ConnectionEdge<T> {
   node: T
   cursor: String
}

generic Connection<T> {
   edges: ConnectionEdge<T>
   pageInfo: PageInfo
}

type Foo {
   id: ID!
   bars: Connection<Bar>
}
```

The overall goal is to reduce the amount of boilerplate in creating schemas with repetitive structures.

**Comment by leebyron** [comment]
Thanks for proposing this! During the GraphQL redesign last year, we actually considered adding this!

There are two issues that seemed to result in more complication than would be worth the value we would get from this feature, but I'm curious what you think or if you have ideas:

1. What does the introspection result look like? When I read the `ConnectionEdge` type and look at its fields, what will the `type` of `node` be? The best answer we could come up with was to change `Field.type` from `Type` to `Type | GenericParameter` which is a bit of a bummer as it makes working with the introspection API more complicated. We could also expand `Type` to include the possibility of defining a generic param itself. Either way, it also has some rippling effects on the difficulty of implementing GraphQL, which would need to track type parameter context throughout most of it's operations.

2. What should `__typename` respond with? What should `{ bars { __typename } }` return? This one is pretty tricky. `{ "bars": { "__typename": "Connection" } }`? That describes the type, but you're missing info about the type parameter, that that ok? `{ "bars": { "__typename": "Connection<Bar>" } }` Is also problematic as now to use the `__typename` field you need to be able to parse it. That also adds some overhead if you were hoping to use it as a lookup key in a list of all the types you know about.

Not to say these problems doom this proposal, but they're pretty challenging.

Another thing we considered is how common type generics would actually be in most GraphQL schema. We struggled to come up with more than just Connections. It seemed like over-generalization to add all this additional complexity into GraphQL just to make writing Connections slightly nicer. I think if there were many other compelling examples that it could motivate revisiting.

**Comment by AndrewIngram** [comment]
You're right about the number of use cases being relatively small, i'll need to think on that point.

To be honest, this feels like sugar for developers of schemas rather than clients. In the simplest case, i'd just expect the introspection result to be the same as it is now, i.e the generics get de-sugared. To that end, it could just be something that parsers of the schema definition language end up supporting, but it's up to library authors how to handle the generated AST.

In graphql-js land, there are numerous examples of libraries ([apollo-server](https://github.com/apollostack/apollo-server), my own [graphql-helpers](https://github.com/depop/graphql-helpers), and a few others I can't remember) which use the parser provided to vastly simplify the process of building schemas (having done it both ways, I'd say it's pretty close to an order of magnitude more productive), and i'd personally be happy to add additional support for tokens related to generics to my library.

However, it does feel weird supporting a syntax that's not actually reflected in the final generated schema, so i'm unsure about this approach.

**Comment by Qard** [comment]
I really wish something like this would be reconsidered. Connections may just be a single use-case, but it's a *big* one, in my opinion. The length of my current schema would cut in half with generics.

Currently I have 24 copies of basically this:

```
type TypeXConnectionEdge {
   node: TypeX
   cursor: String
}
type TypeXConnection {
   edges: TypeXConnectionEdge
   pageInfo: PageInfo
}
```

That's nearly 200 lines of code that could easily be expressed in 8 lines of generics. I'm seriously considering writing my own preprocessor just to hack on my own generics capability...

**Comment by stubailo** [comment]
Hmm, in graphql-tools you could do something like:

```graphql
type MyType {
  hello: String
  world: String
}

${ connectionAndEdgeFor('MyType') }
```

Is there something the syntax could have that would be better than that JS-style approach?

**Comment by Qard** [comment]
And what if you're not using JS? 😞 

I want my schema to be pure graphql schema language so it doesn't *need* preprocessing.

**Comment by stubailo** [comment]
Yeah I definitely sympathize. I guess the real question is, is the generic thing just something for the server to be written more conveniently, or does the client somehow know that these types/fields are generic and acts accordingly?

If the client doesn't know, then I feel like it should be a preprocessor feature or a macro thing. The spec is all about the contract between client and server IMO.

However, there are definitely implementations for generics where the client could actually take advantage of knowledge that something is a generic thing. For example, in the connection example, there's no way to make an interface that says that a `TypeXConnectionEdge` should have a `node` of type `X`, so you can't really enforce that without relying on conventions.

Perhaps this could be done as some sort of intersection of interfaces and type modifiers? So basically, it's a way of creating your own type modifiers - if you squint hard enough, `[ ... ]` and `!` are kind of like `List<T>` and `NonNull<T>`.

So building on that, in introspection:

```
fragment TypeRef on __Type {
  kind
  name
  ofType {
    kind
    name
  }
}
```

You could get:

```
{
  kind: "GENERIC",
  name: "Connection",
  ofType: {
    kind: "OBJECT",
    name: "Photo"
  }
}
```

Perhaps this should come with a change of `kind: "LIST"` to `kind: "GENERIC", name: "LIST"`.

**Comment by Qard** [comment]
I mean, it seems hugely valuable for the client to understand generic concepts too, but that could probably be expressed as simply differently named types, since the client doesn't generally need to be too concerned about the actual name of types so much as what is in them. It seems to me like it'd be really valuable to be able to, in a type-safe way, express concepts like pagination while retaining the safety of recognizing that a given type encapsulates objects of a specific type. Without generics or preprocessing you can sorta-kinda do this with unions, but then you're throwing away the promise that all items in a list are of a specific type...

**Comment by stubailo** [comment]
I guess in my mind, whether or not the client should worry about it is a very important factor in determining whether it should be in the spec or simply live as some server-side library tooling.

**Comment by mike-marcacci** [comment]
Hi everybody! I've definitely been looking to solve the Connection use-case here, and another similar case specific to my project. I have a slightly different strategy which I've laid out in #295, which is a much smaller change, and in no way mutually exclusive with this proposal.

Basically, if an interface were able to explicitly implement another interface, the client would be aware of the hierarchy. That is, it would be provided similar context to what generics might provide, but without adding new semantics to the protocol.

This wouldn't solve the issue of verbosity within the schema, but leverage the existing types to convey the same information. In this way, a preprocessor might be able to compile generics into hierarchal interfaces, achieving both goals here.

**Comment by AndrewIngram** [comment]
Given that we now have at least one authoring-only syntax extension, i.e. `extend type`, is it worth reconsidering generics in the same light?

**Comment by stubailo** [comment]
Hmmm, that's true - `extend type` is not accessible through introspection at all, I didn't think about that.

**Comment by AlecAivazis** [comment]
One use case I have run into for generics is having something resembling `Maybe<T>` to handle error messages. I'd like to do so without having to mess with the network layer and introduce some sort of global state. Currently, I am defining a separate union type for each object (ie `MaybeUser` is a `User | Error`) but it would be nice to be able to do this as simply as `Maybe<User>` and define the structure once. 

An alternative to avoid the extra complexity of the generic union would be something as simple as 
```
generic Maybe<T> {
   left: Error
   right: T
}
```

**Comment by xialvjun** [comment]
Another use case is 
```gql
generic Pagination<T> {
      total: Int!
      limit: Int!
      offset: Int!
      results: [T!]!
}

type Person {
      id
      name
      # ...
}

type Query {
      search_person(limit: Int = 20, offset: Int = 0, q: String): Pagination<Person>!
}
```

**Comment by crypticmind** [comment]
`generic Maybe<T> { ... }` and `generic Pagination<T> { ... }` look great to me, though I'd drop the `generic` keyword as it seems redundant by the use of angle brackets.

**Comment by rasdaniil** [comment]
Addressing problems @leebyron mentioned in the 2nd post here, generics don't need to appear in introspection at all as they are abstract helpers. Concrete types extending generics could be introspected with all the precision of graphql

```graphql
generic Pagination<T> {
    total: Int!
    limit: Int!
    offset: Int!
    results: [T!]!
}

type Person {
    id
    name
    # ...
}

type PaginatedPersons extends Pagination<Person> {
  extraField: Int! // maybe
}

type Query {
    search_person(limit: Int = 20, offset: Int = 0, q: String): PaginationPersons
}
```

In this case the type PaginatedPersons could have in introspection this shape

```graphql
type PaginatedPersons {
    total: Int!
    limit: Int!
    offset: Int!
    results: [Person!]!
    extraField: Int!
}

```

**Comment by dallonf** [comment]
I think there is a potential use case for client-side generics in enabling re-usable components. Example, using a hypothetical react-apollo syntax and the `Pagination<T>` type from the above post:

```
const PaginatedListFragment = gql`
# some purely hypothetical syntax on how you might inject
# another fragment into this one
fragment PaginatedList($itemFragment on T) on Pagination<T> {
  total
  offset
  limit
  results {
    id
    ...$itemFragment
  }
}
`;

const PaginatedList = ({ data, renderItem, onChangeOffset }) => (
  <div>
    {data.results.map(item => <div key={item.id}>{renderItem(item)}</div>)}
    <PaginationControls
      total={data.total}
      offset={data.offset}
      limit={data.limit}
      onChangeOffset={onChangeOffset}
    />
  </div>
);

const PeoplePageQuery = gql`
query PeoplePage($limit: Int, $offset: Int) {
  people(limit: $limit, offset: $offset) {
    # again, hypothetical parametric fragment syntax
    ...PaginatedList($itemFragment: PersonItemFragment)
  }
}
${PaginatedListFragment}

fragment PersonItemFragment on Person {
  id
  name
}
`;

const PeoplePage = ({ data }) => (
  <div>
    <h1>People</h1>
    <PaginatedList
      data={data.people}
      renderItem={item => <a href={`/person/${item.id}`}>{item.name}</a>}
      onChangeOffset={offset => data.setVariables({ offset })}
    />
  </div>
);
]
```

This would be something very powerful for component-based frameworks!

**Comment by AndrewIngram** [comment]
@dallonf this was actually one of the other benefits I envisaged :)

It's the primary reason why I don't think it's enough for this to just be syntactic sugar for the existing capabilities, allowing clients to know about generics could be a very powerful feature.

**Comment by kbrandwijk** [comment]
I also have another use case for generics. I want to define an array of distinct elements, commonly knows as a Set. So instead of defining a field like `myValues: [String]`, I would like to define it as `myValues: Set<String>` to accomplish this.

**Comment by mike-marcacci** [comment]
To add another use case, I've found myself making custom scalars when I would otherwise define a `Map<SomeType>`. In my cases `SomeType` has always been a scalar so this has been acceptable, but there have been other times when I chose to use an array when a `Map<>` would have been preferable.

**Comment by axos88** [comment]
Another use, also tied with pagination is when different interfaces need to be paginated.

Suppose we have an interface `Profile`, and two implementations: `UserProfile` and `TeamProfile`

Without generics, I cannot see a solution to be able to deduct that `UserProfilePagnated` and `TeamProfilePaginated` can be supplied where we are expecting `PaginedProfiles`.
With generics - although not trivial, one can make a complex analyzer that understands the notion of co and contravariance, and can deduct that `PaginatedProfiles` is an interface of `UserPaginatedProfiles`, although one might need to signal this somehow.

**Comment by mike-marcacci** [comment]
Hi @axos88, can you add your comment on #295? Your use case is exactly the kind of problem it is designed to address.

**Comment by axos88** [comment]
@mike-marcacci I'm not sure how interfaces implementing interfaces is enough to have a solution for the problem

**Comment by axos88** [comment]
Although I agree, the problem is related.

**Comment by axos88** [comment]
Oh I see. :)

**Comment by akomm** [comment]
@leebyron

What if generic types can not be returned by fields without type specification. They are just for boilerplate reduction. As you can see in the following definition (from @AndrewIngram's post) the field bars does return the generic type over `T`, but `Bar`:

```gql
generic ConnectionEdge<T> {
   node: T
   cursor: String
}

generic Connection<T> {
   edges: ConnectionEdge<T>
   pageInfo: PageInfo
}

type Foo {
   id: ID!
   bars: Connection<Bar>
}
```

And the `generic` types are not visible in introspection. 

Result of this could be:
```gql
type ConnectionEdgeBar {
   node: Bar
   cursor: String
}

type ConnectionBar {
   edges: ConnectionEdgeBar
   pageInfo: PageInfo
}

type Foo {
   id: ID!
   bars: ConnectionBar
}
```

As far as I know type name duplication is not allowed in graphql schema and all implementation raise errors in case of duplication. So when you already have ConnectionBar somewhere, then you would get error. 

You do not have to drop the information when schema is parsed, so you could point out that ConnectionBar being generated from generic Connection<T> is already defined. In introspection the schema is already processed and the there only exists the ConnectionBar.

**Comment by fbartho** [comment]
@akomm's recommendation is the style that I've gone with in my project.

I currently implemented it with `@generic(...)` as a directive, that I preprocess out to generate the types:

```graphql
interface Page @generic(over: [DataType]) {
	items: [DataType]
	meta: PageMeta
}

type PageMeta {
	page: Int
	totalPages: Int
	totalItems: Int
}

# Elsewhere:

type PopularItems_ implements Page
	@generic(using: [PopularItem_]) {
	searchSectionTitle: String!
}

type PopularItem_ {
	id: ID!
	title: String!
	iconUrl: String
	narrowPhotoUrl: String!
	slug: String!
}
```

After pre-processing, the fields from the Page interface are merged in to the fields of `PopularTemplates_` after distributing the using parameter over all instances of `DataType`

**Comment by akomm** [comment]
So basically graphql need a way to define opaque types/templates to reduce boilerplate, which are not directly exposed as API. For generic results we have interfaces. 

Those templates could have an optional annotation to define how the final type's name is composed - or if it makes sense, a fixed method how it is composed.

**Comment by AndrewIngram** [comment]
As I mentioned in a previous comment, there are reasons why we might not just want this as definition-time syntactic sugar. Exposing the concept of generics to API consumers would allow for highly-reusable client-side code that exploits them. It's currently relatively verbose to implement different connection pagination containers with Relay even though the structure is almost identical every time. Generics are an incredibly powerful construct, so it'd be nice to be able to enhance the GraphQL type system with them.

The main argument against them seems to be that it would increase the difficulty of implementing GraphQL libraries (eg graphql-js, sangria, graphene, absinthe). If the client-side value is limited, i'd agree that it's not worth the complexity. But i'd argue that it's not worth adding at all if it's just a syntactic sugar thing -- there's nothing stopping people implementing these patterns (if slightly less elegantly) in user-land today.

**Comment by alamothe** [comment]
Generics could also be useful to model type-safe IDs:

```graphql
type Person {
  id: ID<Person>!,
  name: String!
}

type Query {
  person(id: ID<Person>!): Person!
}
```

**Comment by akomm** [comment]
@alamothe 

Depends on how you define ID. If it is relay-like opaque string, what would the generic ID<T> type look like?

**Comment by alamothe** [comment]
@akomm It is still opaque to the client. 

Type-safety only prevents the client to pass an ID of the wrong type. For example, client tries to pass `ID<Book>` in a mutation where `ID<Person>` is expected it will get a compile time error

**Comment by akomm** [comment]
@alamothe 
I see

**Comment by AndrewIngram** [comment]
Just adding another use case, where I handle form validation errors in a repetitive way, and having generics would improve client-side reusability:

```
type InvalidField<T> {
  fieldName: T!
  errors: [String!]!
}

type InvalidInput<T> {
  nonFieldErrors: [String!]!
  fieldErrors: [InvalidField<T>!]!
}

enum RegisterUserFieldName {
  NAME
  EMAIL
  PASSWORD
}

type RegisterUserSuccess {
  user: User!
  token: String!
}

union RegisterUserPayload = InvalidInput<RegisterUserFieldName> | RegisterUserSuccess
```

**Comment by crypticmind** [comment]
So, to recap what I see here so far:

The actual approach we're mainly discussing here is adding generics to the language and that would require support from the clients. Some of them will require hacks depending on the language they're coded in.

@akomm mentioned templates, and that's the approach @fbartho used. I could get by with that approach. In the end it reduces boilerplate in the schema definition, and clients require no change. The only downside I see is that clients may be exposed to many identical classes with different names, which can be cumbersome when coding.

@leebyron could you tell me how the language's RPF process works, and what information would be required to actually push forward this proposal here?

**Comment by leebyron** [comment]
I'm still concerned that this adds a lot of additional complexity to GraphQL libraries and client tools. I definitely understand and see the potential value, but I think it should be made very clear that the added value is well worth the complexity cost.

Next steps to do so would be to implement this as a PR in GraphQL.js to better understand complexity, and write the spec text to understand changes to schema and query validation algorithms

**Comment by robbyemmert** [comment]
I think there's a fundamental underlying conversation that seems to be going on here, and I've had it with a few developers now:

GraphQL aims to be "a complete and understandable description of the data in your API" (taken from the graphql website homepage). Graphql also aims to be simple. These two seem to directly collide when you have a complex API (any enterprise-level API really). But I think the answer to this depends on the angle of approach. (And the angle will likely depend on your team and implementation).

Here are two angles of approach to using GraphQL:
**Approach A**
1. Write my schema in GraphQL first (could be by a front end dev, back end dev, or even a Product Manager for all we care)
2. Create (or generate, if possible) native types and resolvers that implement the schema.

**Approach B**
1. Write my business logic in my language of choice.
2. Generate GraphQL schema from the native types I'm already using, much as I would generate REST documentation for something like swagger.

Personally, I think both approaches are valid. However, I think that when discussing more complex syntax, it's worth mentioning that Approach A requires stronger features like Generics, whereas Approach B does not.

GraphQL (and the GraphQL community **ahem** Graphcool/Prisma) both heavily advertise and encourage Approach A. Approach A also has some great pros for highly collaborative environments. The paradigm of having a unified language in which to describe web services that is concise and can be understood by the whole team is very lucrative for many organizations. Hence you see phrases like "a complete and understandable description of the data in your API" on the GraphQL website.

In my opinion, if GraphQL is going to be a long-term viable option for approach A, it needs to be able to describe complex environments, which means supporting things like generics and (:trollface:) custom directives. 

I love where the conversation is going, I just wanted to point out that complexity isn't necessarily a bad thing in this case, and really, I think both approaches can be supported in the long run. This is ultimately what gives GraphQL the potential to become an industry standard, and what differentiates it from things like Swagger.

My TLDR 2 cents.

**Comment by leebyron** [comment]
Excellent analysis!

I agree that GraphQL (intentionally) tries to fulfill competing ideals of fully describing your data and being as simple as possible. That often forces compromise. 

Also I like your framing of how to build schema, but it lacks a third approach I’ll call “C” which is what most are doing today: First write my business logic in my language of choice, then manually write a GraphQL schema in that same language.

While I agree all approaches are valid, out of the three I strongly prefer your “approach B”. It turns out that programming languages are great for describing business logic and GraphQL isn’t a programming language. In cases where your programming language of choice comes with a type system as or more capable than GraphQL’s then generating it the same way you might generate a Swagger schema is ideal. In fact: generate both GraphQL and Swagger and have multiple API facades to the exact same business logic!

**Comment by dallonf** [comment]
I'd also like to add my perspective on the issue, which is that generics in GraphQL would be largely a client-side benefit rather than server-side (although it could certainly be that as well), allowing for much more flexible fragments. 

Since query generation on the client side seems to be strongly discouraged (which I at least partly understand as wanting to keep the queries explicit and easy to read), there's really no other way to enable things like a pagination component fragment as I mentioned above (https://github.com/facebook/graphql/issues/190#issuecomment-343487643).

**Comment by leebyron** [comment]
> I worry about approach B, as it tends to leak internal implementation details without producing a holistically designed API

I agree that poorly developed this could definitely occur. But also a naive translation of existing business logic to a manually written GraphQL schema would have the same outcome. To be successful, the parts of the internal logic exposed to GraphQL needs to be intensional and designed with API exposure in mind. This advice applies equally to exposing a REST API with generated Swagger

**Comment by joe-io** [comment]
I think there is a long history in APIs of having clean, simple IDLs.  In fact, that is what attracts a lot of people to GraphQL in the first-place.

It is also what attracts a lot of people to gRPC / Protocol Buffers.

In a polyglot, multi-team world, I want the ability to have a clean, language-agnostic interface description of my API.

Not only does it allow tooling support, but more importantly it allows the humans who write the software in different languages, tools, and teams, to come together and agree in a shared way on the API.

To me, is one of the strongest reasons why I choose to use GraphQL.

**Comment by robbyemmert** [comment]
> I think there is a long history in APIs of having clean, simple IDLs. In fact, that is what attracts a lot of people to GraphQL in the first-place.
> 
> It is also what attracts a lot of people to gRPC / Protocol Buffers.
> 
> In a polyglot, multi-team world, I want the ability to have a clean, language-agnostic interface description of my API.
> 
> Not only does it allow tooling support, but more importantly it allows the humans who write the software in different languages, tools, and teams, to come together and agree in a shared way on the API.
> 
> This is one of the strongest reasons why I choose to use GraphQL.

I totally agree. That's why we need a clean, simple way to describe generic types.

**Comment by thetrompf** [comment]
What if we only have generic types on "compile time" so they don't exist in the runtime, then we only need something lige a "template expansion mechanism" which "compiles" to concrete types.

```gql
template Connection<TEdge> {
    edges: [TEdge]!
}

template Edge<T> {
    node: T
}

type ProfileEdge = Edge<Profile>
type ProfileConnection = Connection<ProfileEdge>

type Profile {
    id: ID!
}

type Query {
    profiles: ProfileConnection!
} 
```

You can then implement higher order functions to deal with the "generic types" and call them in the resolvers on specific fields.

**Comment by AlecAivazis** [comment]
@thetrompf have you seen [graphql-s2s](https://github.com/nicolasdao/graphql-s2s)? It basically does exactly what you're suggesting.

However, not exposing generics to the client prevents a whole family of components built around recurring structures in an API from being possible. For example, a component that knows how to render a paginated list with page numbers. Not to mention it would mean that a subset of the SDL only applies in a particular context which I'm not sure is appropriate.

**Comment by thetrompf** [comment]
@AlecAivazis yeah I know of the project, and you're right. s2s compiles to GraphQL SDL compliant schemas. But it would be great if the official languages supported this. s2s does not play well if you want to combine it with other tools operating on the schema like graphql-code-generator.com, so I have to defined a pre-compile step before invoking the code generator, because I'm now working with a superset of the SDL.

**Comment by thetrompf** [comment]
Yeah I can see there are shortcomings with this solution.
The paginated list problem could be solved with interfaces though.

Med venlig hilsen / Best regards
Brian Kejlberg

ons. 5. jun. 2019 11.27 skrev Alec Aivazis <notifications@github.com>:

**Comment by mike-marcacci** [comment]
@thetrompf paginated can _almost_ be solved with interfaces... and can totally be solved once #373 lands.

**Comment by Avi-D-coder** [comment]
The longer this proposal languishes, the more bad duplicated code is written.
We can think of generics as adding complexity to the spec and implementations or we can think of them as reducing complexity for graphql's users. Given that just about every modern typed language has or will soon have generics (cough go), the choice is clear as day.


For generics at runtime to be most useful you need to be able to talk about `T`.
`Ord<T>` should mean `T implements Ord` where `Ord` is defined as:
```graphql
interface Ord<T> {
  compare(a: T!, b: T!): Ordering!
}

enum Ordering {
  GT
  EQ
  LT
}
```

**Comment by Avi-D-coder** [comment]
In fairy dreamland where graphql's type system is powerful enough to define actual query languages.



```graphql
query {
  compare(droid(id: "2000"){name}, hero(episode: EMPIRE) {name})
}
```

This would depend on both sets fields having the same types and all those types implementing `Ord`. 
In other words, duck typing plus higher order fields would allow graphql to contain a query language like SQL.

**Comment by brandonbyskov** [comment]
Generics would improve the readability of GraphQL APIs.

From the [Shopify graphql design tutorial](https://github.com/Shopify/graphql-design-tutorial) that I really like, the one part that bothers me is the [return types suggested for mutations](https://github.com/Shopify/graphql-design-tutorial/blob/master/TUTORIAL.md#output):
```graphql
type CollectionCreatePayload {
  userErrors: [UserError!]!
  collection: Collection
}
```

Having to create a new type for every mutation and requiring the user look up that type to know what it is seems excessive when you could simplify types with:

```graphql
type CreatePayload<T> {
  userErrors: [UserError!]!
  result: T
}
```
And we could use the `CreatePayload<Collection>` return type instead.

**Comment by alamothe** [comment]
GraphQL is pretty much dead at this point — no progress is being made for years. Same situation is with other RFCs (input union, empty types).

**Comment by mike-marcacci** [comment]
@alamothe when your message hit my inbox I genuinely thought it was satirical and laughed a bit. Whether it was serious or not, it’s pretty darn far from the truth. This year the GraphQL project has migrated to its own foundation and has really stepped up its pace after an admittedly long silence. The project has a strong and committed leadership team which has been making steady progress on a lot of these concerns. If you’d like to track the progress of these bigger conceptual changes, check out the notes from the monthly working group meetings – or better yet, if you have something to contribute, join one! If you aren’t able to be that involved but want to track progress, subscribe to this repo: the two issues you mentioned are both under active consideration.

**Comment by akomm** [comment]
Also worth mentioning is:

while this problems are being worked on, I worked with the graphql server implementation on two different language and all of them had basically an out of the box capability to reduce the boilerplate and make semi-generics possible. For example, given the JS implementation if you use the object form of schema creation you can simply make factory functions that construct variants of a type based on input. Using typescript on top of this and you basically have what you want without having to implement additional abstraction layer, just using the graphql server library code and making factory functions.

This kind of mitigated the pressure & need for the feature to me personally. If there is a good solution for this in-language one day, great. If not, then not a big deal and definitely not the death of graphql.

**Comment by alamothe** [comment]
I hope things will change, but I'm not overly optimistic. Each of these RFCs was opened a few years back.  Thanks for the reply!

**Comment by wtrocki** [comment]
Based on quick investigation I think could be a couple of approaches to resolving the problem with the type name

1) Preprocessor/Templating

Having extra type defined in the parsed schema object that is not defined in the SDL. This brings a number of challenges and probably it is not suitable for the spec implementation as it forces developers to use types that are not defined in SDL. From my point of view this could be the way how the community can implement generics now without involving changes in the spec.

An example was mentioned above:
https://github.com/graphql/graphql-spec/issues/190#issuecomment-287686157

Pure hack for syntax using existing spec could look like this

```graphql

type Page @generic {
  offset: Number
  total: Number
  """ items: [GenericField] """
}

type User @createGenericArray(name: "items", type: "Page") {
  id: ID
}

type Query {
  getUsers: UserPage
}
```
After processing 
```
 type User {
  id: ID
 }

type  UserPage {
  offset: Number
  total: Number
  items: [User]
}

type Query {
  getUsers: UserPage
}
```


This is a pure hack but I have used this to evaluate options and see some challenges. Result schema:

- Missing Page type
- Contain `UserPage` type
- Devs need to use `UserPage` type without being defined :(

2) Explicit type definitions for generics 

Any usage of generics could require an explicit definition of the type that will be represented. 

Explained already in a couple of comments in this issue
https://github.com/graphql/graphql-spec/issues/190#issuecomment-343376473
and
https://github.com/graphql/graphql-spec/issues/190#issuecomment-498971680

## Summary

For some of the user cases, code generation can be employed to add some form of templating capabilities to schema and generate schema that contains types explicitly defined (For example https://www.opencrud.org). This is a pretty well-known workaround for the moment. 

As for the spec having the ability to define explicit type for generics could resolve this problem on the spec level - although it could look too verbose.

**Comment by justinfagnani** [comment]
I'm just learning GraphQL, but already I'm seeing the need for generic types in my use cases. In particular I need to deal with a Reference type in my backend, where the value can reference any other type. There can be generic fetch operations that can retrieve the value of a ref.

Something like so:

```gql
type User {
  manager: Ref<User>;
}

type Query {
  getRef(ref: Ref<T>): T
}
```

**Comment by benjie** [comment]
@justinfagnani That seems like a use case for interfaces or unions.

**Comment by xialvjun** [comment]
I don't think it's a tech problem for not implement this for 5 years. 
It's just there is no one can decide should we make it.

**Comment by dncrews** [comment]
Just to add a use-case that I've been struggling with, which is the "other side" of "discoverable query successes" (Connections) is "discoverable mutation failures":

### Mutation Errors
_In the schema_, we're trying to call out all of the potential failure cases when they're "the user can do something to fix this, it's not a critical failure". To do this, we're introducing an error `code` that is an enum rather than a string. This one change makes what would've been a shared generic response type into triple the amount of schema. It feels wrong that making my schema self-documenting makes it _significantly_ noisier.

Here's what I wish I could do:

```graphql
interface MutationError<CodeEnum> {
  path: [String!]
  message: String!
  code: <CodeEnum>!
}

enum USER_CREATE_ERROR_CODE {
  USERNAME_TAKEN
  PASSWORD_REQUIREMENTS_NOT_MET
  …
}

interface MutationFailure<CodeEnum> {
  errors: [MutationError<CodeEnum>!]!
}

type UserCreatePayload = UserCreateSuccess | MutationFailure<USER_CREATE_ERROR_CODE>

type Mutation {
  userCreate(…): UserCreatePayload!
}
```

With the request
```graphql
mutation CreateUser {
  userCreate(…) {
    ... on UserCreateSuccess {
      ... UserInfo
    }
    ... on UserCreateFailure {
      errors {
        path
        message
        code
      }
    }
  }
}
```

Adding extra pain to this is [this related enum issue](https://github.com/graphql/graphql-spec/issues/496), so I can't even do this with interfaces at all, as you'd expect.

**Comment by akomm** [comment]
GraphQL libraries in different languages usually provide at least two methods of defining the schema. 
1. Using graphql semantics to write the types and a resolver map that maps fields
2. Using some sort of alternative configuration format, like yaml or write code to create schema

I'd prefer the first method, but actually use the latter one. The reason hereby is the lack of abstraction/generics. With code solution I can write a factory that creates a non-generic type using arguments representing generics. This way I can abstract. 

Would be nice if I could use the 1. method without losing the abstraction

**Comment by fluidsonic** [comment]
We also lack generics in various situations, like this one:

```graphql
type Change<Value> {
   old: Value
   new: Value
}

type Contact {
   companyName: String
   firstName: String!
   id: ID!
   lastName: String!
}

type ContactUpdatedJournalItem {
   companyName: Change<String>
   firstName: Change<String!>
   id: ID!
   lastName: Change<String!>
   // etc.
}
```

We commonly use this pattern to create journal items for an object's change history.
- For each property of the object we log a change in its `Change<T>` property.
- If there is no change, then `Change<T>` is `null`.
- Changes on a nullable `T` can be `{ old: "foo", new: null }`, `{ old: null, new: "foo" }` and `{ old: "foo", new: "bar" }`.
- Changes on a non-nullable `T!` can only be `{ old: "foo", new: "bar" }`.

For us it's important that the use-site of the generic type specifies if `T` is nullable or not.

But there may be other scenarios. The question is what are the possible combinations and use cases?

```graphql
type Foo<T> {
   t1: T      // String or String! for Foo<String!>  ??
   t2: T!     // String!
   t3: T?     // String   (if nullability of T can be stripped)
}

type Bar {
   foo1: Foo<String>
   foo2: Foo<String!>
}
```

And should it be possible to force non-null types?

```graphql
type Foo<T!> { // must be non-nullable
   t1: T      // String, String!, or forbidden  ??
   t2: T!     // String!
}

type Bar {
   foo1: Foo<String>   // not allowed
   foo2: Foo<String!>  // ok
}
```


Regarding other generic scenarios, similar to edges, we have something like this in Kotlin:
```kt
class SearchResult<out Element : Any>(
    val elements: List<Element>,
    val nextOffset: Int?,
    val totalCount: Int,
)

class SearchQueryOutput(
    val cityResult: SearchResult<City>?,
    val contactResult: SearchResult<Contact>?,
    val locationResult: SearchResult<Location>?,
    val processResult: SearchResult<Process>?,
)
```

Our own Kotlin GraphQL library translates it to something like this:

```graphql
type CitySearchResult {
    elements: [City]!
    nextOffset: Int
    totalCount: Int!
}

type ContactSearchResult {
    elements: [Contact]!
    nextOffset: Int
    totalCount: Int!
}

type LocationSearchResult {
    elements: [Location]!
    nextOffset: Int
    totalCount: Int!
}

type ProcessSearchResult {
    elements: [Process]!
    nextOffset: Int
    totalCount: Int!
}

type SearchQueryOutput {
    cityResult: CitySearchResult
    contactResult: ContactSearchResult
    locationResult: LocationSearchResult
    processResult: ProcessSearchResult
}
```

The transformation automatically specializes generic types. That quickly blows up the GraphQL type system and makes it quite cumbersome to use on the client side. It also makes the server's GraphQL library implementation more complicated.

**Comment by fluidsonic** [comment]
Regarding the templating approach it quickly gets out of hand when Lists and NonNull types are involved.
From my example above, there might be:
- `Change<String>`
- `Change<String!>`
- `Change<[String]>`
- `Change<[String!]>`
- `Change<[String]!>`
- `Change<[String!]!>`

Which would expand to an increasingly weird type system, e.g.:
```graphql
type StringChange { … }
type NonNullStringChange { … }
type NonNullStringChange { … }
type StringListChange { … }
type NonNullStringListChange { … }
type StringNonNullListChange { … }
type NonNullStringNonNullListChange { … }
```
(add more List nesting for more fun)

Another question: Would it be allowed to use generic types as a generic argument?
```graphql
type Foo<T> {
   foo: T
}

type Bar<T> {
   bar: T
}

type Baz {
   baz: Bar<Foo<String>>
}

// { baz: { bar: { foo: "xyz" } } }
```

**Comment by n614cd** [comment]
In many ways GraphQl lends itself to data driven architectures; except the lack of generics, it pretty much kills the ability to use a common set of queries and mutations. Technically, you can almost solve this with interfaces.

In my current system, I have roughly fifty entities, which are flattened down to twenty five objects consumed by the UI. The effectively means instead of four query/mutations for manage and perform all CRUD, I have a hundred. (search, create, update and delete per UI object). This makes the API very unwieldy, and is before we get into any transactional specific APIs.

Tim

**Comment by dan-turner** [comment]
Any progress or developments on this?

**Comment by IRod22** [comment]
I would recommend going off of the TS Spec because it has almost everything GraphQL needs for generics. 

Regarding implementation, I think I have an idea for a new typed API. Here's how it would look in JavaScript:

```js
const Change = GQLAlias('Change', ['T'], T => GQLObject([
    GQLProperty('old', T.NotNull()),
    GQLProperty('new', T.NotNull()),
]))

// usage:
const schema = Schema([
    Property('getLatestChange', Change(GQLString)),
    // ...
])
```

That is just an idea; I am going to try to implement a better schema/resolver API for a smaller library to test this example.

Also, I recommend replacing the `!` operator with a sort of `Maybe<T>` for better type safety.

**Comment by rgolea** [comment]
I can't wait for this to become a reality. It's the main reason I had to move away from graphql as it become so hard to maintain and work with. I had to remember all these properties I had to distribute everywhere. And yes, implements works pretty well but at the end of the day I had to copy all the properties around.

**Comment by akomm** [comment]
This is actually reality all the time. But people decided going schema-first and gave away the abstraction of code-first. Without the abstraction they have to repeat everything over and over again.

You can always use functions to create new types in code, considering your function params as type params. You can even create factory functions so at the end you can reason about your schema (in code) similar to the schema in the graphql DSL.

**Comment by DeveloperTheExplorer** [comment]
It has been almost 7 years. Any updates? This is very much needed for schema/resolver definitions. Almost all endpoints need this whenever pagination is involved. And almost all systems implement some sort of pagination/cursor.

**Comment by akomm** [comment]
Going schema DSL in the graphql implementations is one of the biggest mistakes IMO. You can see it by all those "solutions" cooked around it by now, to fix problem A, B, C and problems still being present to this day. Whoever stayed on the code side without all the mess around it, has much less problems with graphql.  

Like always, I always try out new tech and evaluate it. It feels tempting at first glance, but the price vs. benefit is hugely disproportional. 

You can have quite clean schema definitions also in code, without all the primitive constructors provided by the impl. libraries. And you can eliminate repetition that you try to solve with generics using code.

The examples I've seen so far here, that should prove the code approach without generics leads to bad naming, feels artificial. Most of the time its a matter of `Change<String>` vs. `ChangeString`. The example variants with nullability are not really useful. To talk about whether its a problem or not needs some real world example to see whether its even needed for the type in the specific case to have the nullability encoded in the name this way. I can just imagine RL examples for that where you don't actually need it, unless you make some weird design decision. If you want to avoid collision in name, the first question is why do you have those two variants, what is the intent of the data structure you try to describe and isn't the nullability something that is rather implied from a different type name as a base (OptionalChange<T> vs Change<T>) instead. 

I'm not saying there is no problem, but just that the examples I've seen here so far are IMO to artificial to be convincing.

**Comment by jamietdavidson** [comment]
Would like to reup this. Seems relevant and I stumble into Enum / Interface related issues every few months that would be solved by this. It's kind of the last missing piece to making GraphQl not leaving something to be desired, IMO.

**Comment by n614cd** [comment]
Is there a specific blocker to making this happen?
Resources to push it through? Some unsolved problem?

**Comment by varvay** [comment]
At first I was excited to the idea about generic type support in GraphQL, since it might introduces high degree of Intuitiveness for developer to translate the GraphQL schema into code implementation in various language and framework. It's the same concept as having scalar data types as much as possible to match all the possible implementor languages e.g., think how intuitive is it to implement the schema when it supports data type like hashmap, JSON node, etc.

But I've been thinking to myself about this and end up by accepting the current GraphQL specification without generic type support. The question i've been asking myself are,

* are the issue regarding the need of generic type should be handled by the specification or the implementation (code implementation and framework)?

So I started from the mindset that GraphQL is a specification used as contract between frontend and backend on what data they will exchanging and how are they structured. There will be complication introduced with generic type implementation on these information received by the client, for example how does the client knows the structure of the object defined as generic? There must be an information to communicate it right? you'll ended up by sending some kind of metadata, which is redundant to `__typename`.

The next question is,

* how generic type support in the specification helps us in the system design and implementation process? Is it by increasing the quality of the information delivered to the client? Is it decreasing the complexity of these process significantly?

I don't think so. The client's frameworks will still need to implement the data resolution abstraction and the quality of the information received by the client doesn't necessarily increased since the only additional information are "this object is generic and it might be in type A, B or C", which is communicated already through the usage of union type definition.

* can't the current specification fulfill the abstract needs for generic types e.g., pagination and error-messaging?

I think it does fulfill them with two possible solutions on the table,
1. using interface.
2. using union data type, which is the one I prefer.

Discussing the 2nd solution further,
```
type Post {
   # some type definition
}

type Profile {
   # some type definition
}

type Tag {
   # some type definition
}

type Setting {
   # some type definition
}

union Pageable = Post | Profile | Tag | Setting # Verbosity potential

type PagedResult {
   data: [Pageable]
   page: Int
   # some other pagination related definition
}

type Query {
   getPosts(): PagedResult
   getProfiles(): PagedResult
   getTags(): PagedResult
   getSettings(): PagedResult
}
```

Such query is resolvable in the code implementation by mapping the type based on the `__typename` metadata, for example like using discriminated union in typescript. I also still consider the schema definition verbosity is acceptable since the only verbosity came from `Pageable` union type definition, since for new type extending pagination functionality will goes to this list. But the tradeoff with strong data typing and information quality build upon the contract are reasonable.

Or what you really asking is some kind of no-type data and omitting the type enforcement feature? I think this is oppose to the main goal of GraphQL itself.

Finally, based on those thinking, I concluded for myself that in use cases I've seen so far, the only benefit I'm going after is syntactic sugar. This might also be your case. In my opinion, If there are any work should be done regarding this, they should be on the client side (developer and implementor framework).

**Comment by mathroc** [comment]
> for example how does the client knows the structure of the object defined as generic? There must be an information to communicate it right? you'll ended up by sending some kind of metadata, which is redundant to __typename.

no, if you have:

```gql
type Post {
  name: String!
}

type List<T> {
  nodes: [T!]!
}

type Query {
  posts: List<Post>
}
```

then there is no doubt about what `posts.nodes` contains in `query { posts { nodes {}}}`

> the only additional information are "this object is generic and it might be in type A, B or C"

This is not what generics do

and it's probably the same misunderstanding, but if you use generics in your example, it becomes:

```gql
type Post {
   # some type definition
}

type Profile {
   # some type definition
}

type Tag {
   # some type definition
}

type Setting {
   # some type definition
}

type PagedResult<T> {
   data: [T]
   page: Int
   # some other pagination related definition
}

type Query {
   getPosts(): PagedResult<Post>
   getProfiles(): PagedResult<Profile>
   getTags(): Paged<Tag>
   getSettings(): PagedResult<Setting>
}
```

And there's definitely readability improvements as-well as better typings than in the previous version (you know that `query.getPosts.data `will always be of type `[Post]`, not `[Post|Profile|Tag|Setting]`

**Comment by nort3x** [comment]
@mathroc 

inspired by how C++ resolve templates, we can preprocess schema and produce this pipe:

input schema:
```graphql

type PagedResult<T> {
   data: [T]
   page: Int
}

type Query {
   getPosts(): PagedResult<Post>
   getProfiles(): PagedResult<Profile>
}
```

processed schema:

```graphql
type PagedResultPost{
   data: [Post]
   page: Int
}

type PagedResultProfile{
   data: [Profile]
   page: Int
}

type Query {
   getPosts(): PagedResultPost
   getProfiles(): PagedResultProfile
}
```

notice that we can't really mangle the outcome because it should be follow-able by consumer of the API  
another issue is that the consumer should also use the same processor or re-implement them one by one

i used this in a small passion project and wasn't really an issue, but i understand the complication and integrity problems it could bring into the specification...  

8 years is alot i think graphql specification should stay as is and task these improvements to it's successors

**Comment by aradalvand** [comment]
Zero progress on such an essential lack of a feature lack for 9 years? How does this make sense? One would think GraphQL is dead judging by the (near-zero) concrete progress on the most popular proposals in this repo, including but not limited to this one.
